internal/servers/worker/internal/metric/proxy_ws.go

```diff
@@ -0,0 +1,127 @@
+package metric
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	proxyWebsocketSubsystem = "worker_proxy_websocket"
+)
+
+var (
+	activeConnCount = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "active_connections",
+			Help:      "Count of open websocket proxy connections (to Boundary workers).",
+		},
+	)
+
+	bytesReceived = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "received_bytes_total",
+			Help:      "Count of received bytes for Worker proxy websocket connections.",
+		},
+	)
+
+	bytesSent = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "sent_bytes_total",
+			Help:      "Count of sent bytes for Worker proxy websocket connections.",
+		},
+	)
+)
+
+// InitializeWebsocketCollectors registers the websocket collectors onto `r`.
+// It panics upon the first registration that causes an error.
+func InitializeWebsocketCollectors(r prometheus.Registerer) {
+	if r == nil {
+		return
+	}
+	r.MustRegister(activeConnCount, bytesReceived, bytesSent)
+}
+
+// InstrumentWebsocketWrapper expects an http.Handler where websockets are used
+// and wraps it to instrument that websocket with metrics (bytes received, sent, etc).
+func InstrumentWebsocketWrapper(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		nextRw := rw
+		if rwh, ok := rw.(http.Hijacker); ok {
+			nextRw = &hijackerWrapper{ResponseWriter: rw, Hijacker: rwh}
+		}
+		next.ServeHTTP(nextRw, req)
+	})
+}
+
+type wsConn struct {
+	*sync.Once
+	net.Conn
+}
+
+func (wc *wsConn) Read(b []byte) (int, error) {
+	n, err := wc.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+	bytesReceived.Add(float64(n))
+	return n, nil
+}
+
+func (wc *wsConn) Write(b []byte) (int, error) {
+	n, err := wc.Conn.Write(b)
+	if err != nil {
+		return n, err
+	}
+	bytesSent.Add(float64(n))
+	return n, nil
+}
+
+func (wc *wsConn) Close() error {
+	wc.Once.Do(func() { activeConnCount.Dec() })
+	return wc.Conn.Close()
+}
+
+type hijackerWrapper struct {
+	// The `Once` objects constrain incrementing/decrementing active connections.
+	// Only a precautionary measure as:
+	// 1. http.(Hijacker).Hijack() cannot be called more than once (2nd call is an error).
+	// 2. websocket.(*Conn).Close() cannot be called more than once (2nd call is no-op).
+	inc sync.Once
+	dec sync.Once
+
+	// `ServeHTTP` takes an `http.ResponseWriter`, so `metricsWrapper`
+	// needs to implement it. Other than that, `ResponseWriter` is unused.
+	http.ResponseWriter
```

Comment:
I'm not familiar with this pattern, but we only wrap `http.ResponseWriter` if it is `http.Hijacker`

Reply:
The thing is that we wrap the http.Hijacker interface only so if any downstream code relies on http.Pusher or http.Flusher, the ResponseWriter we pass onto it will never satisfy those interfaces.  Happy to discuss this further.

Reply:
Ahhh I see - got it now, thanks for explaining 👍